<br>

# DSG ASSIGNMENT


<BR>

In [1]:
#importing important modules and libraries

#to plot on the notebook itself
%matplotlib inline 


import numpy as np
import scipy as sp
import matplotlib as mlp
import matplotlib.cm as cm
import matplotlib.pyplot as plt
import pandas as pd

#CUSTOMISING DISPLAY OPTIONS
pd.set_option('display.width', 500)
pd.set_option('display.max_columns', 100)
pd.set_option('display.notebook_repr_html', True)
import seaborn as sns

In [2]:
#loading the test data and train data on jupyter notebook
test_original=pd.read_csv('/Users/anurag/Downloads/ml/DSG /DSG_Predictive Modelling/mushroom_test.csv')
data=pd.read_csv('/Users/anurag/Downloads/ml/DSG /DSG_Predictive Modelling/mushroom_train.csv')
data.head(5)


class cap-shape cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size gill-color stalk-shape stalk-root  radius stalk-surface-above-ring stalk-surface-below-ring stalk-color-above-ring stalk-color-below-ring veil-type veil-color  weight ring-number ring-type spore-print-color population habitat
0     p         f           s         e       f    s               f            c         n          b           t          ?     215                        s                        k                      p                      w         p          w      38           o         e                 w          v       l
1     p         x           y         w       t    p               f            c         n          k           e          e      94                        s                        s                      w                      w         p          w      43           o         p                 n          v       u
2     e         f           s         n       t    n               f            c         b          e           e          ?     244                        s                        s                      e                      e         p          w      37           t         e                 w          c       w
3     p         x           s         e       f    y               f            c         n          b           t          ?     286                        k                        s                      p                      w         p          w      82           o         e                 w          v       l
4     e         x           s         w       t    l               f            c         b          n           e          c     110                        s                        s                      w                      w         p          w      36           o         p                 n          s       m

# Analysing the Data

In [3]:
data.shape

(5686, 25)

So we have 5686 given training sets with 24 features.
THe data describes the edible property of mushroom species (whether poisonous[p] or edible[e]).
Notice the column with heading 'stalk-root'

In [4]:
data['stalk-root'].head()

0    ?
1    e
2    ?
3    ?
4    c
Name: stalk-root, dtype: object

In [5]:
print "No. of unkown data= "
a= np.sum(data['stalk-root'] =='?')
print a

No. of unkown data= 
1727


A lot of data is not known in stalk-root. 
#One way to procced is to give '?' as one of output type while labelling 
#OR
#First predict stalk-root from the remaninig training sets before actualling modelling for preddiction of class

Let's see now how many we have continous variables and how many discreet variables


In [6]:
print data.dtypes
data.shape[0], data.shape[1]

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
radius                       int64
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
weight                       int64
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object


(5686, 25)

 From the above analysis only weight and radius of the species are continous variables

<br>

# Categorising object variables

This is required since we need to process the data from object to more readbale form of integer values only.

<br>

In [7]:
categorical_variable=data.dtypes[data.dtypes=='object'].index.values

In [8]:
categorical_variable

array(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'], dtype=object)

In [9]:
#Find number of unique outputs for every variable (independent as well as dependent variable)
data[categorical_variable].apply(lambda x: len(x.unique()))

class                        2
cap-shape                    6
cap-surface                  4
cap-color                   10
bruises                      2
odor                         9
gill-attachment              2
gill-spacing                 2
gill-size                    2
gill-color                  12
stalk-shape                  2
stalk-root                   5
stalk-surface-above-ring     4
stalk-surface-below-ring     4
stalk-color-above-ring       9
stalk-color-below-ring       9
veil-type                    1
veil-color                   4
ring-number                  3
ring-type                    5
spore-print-color            9
population                   6
habitat                      7
dtype: int64

It can be seen that 'veil-type' has one type only. So it has no effect on model.


# Preprocessing data

<br>

In [10]:
#describing independent and dependent variables
dependent_variable= 'class'
independent_variable= [ x for x in data.columns if x not in [dependent_variable]]

In [11]:
#changing the objects (stored as categorical variable) to descreet classes.

from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

for var in categorical_variable:
    data[var]=le.fit_transform(data[var])

In [12]:
data.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  gill-spacing  gill-size  gill-color  stalk-shape  stalk-root  radius  stalk-surface-above-ring  stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  weight  ring-number  ring-type  spore-print-color  population  habitat
0      1          2            2          2        0     7                1             0          1           0            1           0     215                         2                         1                       6                       7          0           2      38            1          0                  7           4        2
1      1          5            3          8        1     6                1             0          1           4            0           3      94                         2                         2                       7                       7          0           2      43            1          4                  3           4        5
2      0          2            2          4        1     5                1             0          0           1            0           0     244                         2                         2                       2                       2          0           2      37            2          0                  7           1        6
3      1          5            2          2        0     8                1             0          1           0            1           0     286                         1                         2                       6                       7          0           2      82            1          0                  7           4        2
4      0          5            2          8        1     3                1             0          0           5            0           2     110                         2                         2                       7                       7          0           2      36            1          4                  3           3        3

Notice that stalk-root with '?' has been changed to value of  '0'. 


<br>

# Training model without predicting unknown stalk-root data

<br>

<br>

In [13]:
#splitting training data in test, train, test size is 20% of whole train data
from sklearn.model_selection import train_test_split
train, test = train_test_split(data, test_size=0.2, stratify=data['class'])

In [14]:
print(len(data))
print(len(train))
print(len(test))

5686
4548
1138


# Logistic Regression Model

<br>

In [15]:
#importing Logistic Regression Algorithm
from sklearn.linear_model import LogisticRegression

In [16]:
#creating our model
model1=LogisticRegression()

In [17]:
#applying Logistic Regression on train data
model1.fit(train[independent_variable], train[dependent_variable])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [18]:
#storing predicted test['class']
prediction1= model1.predict(test[independent_variable])

In [19]:
prediction1

array([0, 0, 1, ..., 0, 1, 0])

# Accuracy score

In [20]:

from sklearn.metrics import accuracy_score 
accuracy_score(prediction1, test['class'])

0.94639718804920914

#  Evaluation using confusion matrix

In [21]:
from sklearn.metrics import confusion_matrix
confusion_matrix(test['class'], prediction1 )

array([[565,  24],
       [ 37, 512]])

Note: The confusion matrix here describes how many correct predictions and wrong predictions have been made <br>
    for class=0 and class=1. 
    Total wrong predictions=34+22=56

# Classification Report

In [22]:
from sklearn.metrics import classification_report
print classification_report(test['class'], prediction1 )

             precision    recall  f1-score   support

          0       0.94      0.96      0.95       589
          1       0.96      0.93      0.94       549

avg / total       0.95      0.95      0.95      1138



<br>

<br>

# Decision Tree Model

<br>

In [23]:
#inporting Decision Tree Algorithm
from sklearn.tree import DecisionTreeClassifier

#creating our model
model2=DecisionTreeClassifier()




In [24]:
#applying DECISION TREE  on train data
model2.fit(train[independent_variable], train[dependent_variable])



DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [25]:
#storing predicted test['class']
prediction2=model2.predict(test[independent_variable])
prediction2

array([0, 0, 1, ..., 0, 1, 0])

# Accuracy score



In [26]:
accuracy_score(prediction2, test['class'])

1.0

#  Evaluation using confusion matrix


In [27]:
confusion_matrix(test['class'], prediction2 )

array([[589,   0],
       [  0, 549]])

All the predictions are true

# Classification Report

In [28]:
print classification_report(test['class'], prediction2 )

             precision    recall  f1-score   support

          0       1.00      1.00      1.00       589
          1       1.00      1.00      1.00       549

avg / total       1.00      1.00      1.00      1138



<br>

# COMPLETING ORIGINAL TEST DATA

In [29]:
test_original.head()

cap-shape cap-surface cap-color bruises odor gill-attachment gill-spacing gill-size gill-color stalk-shape stalk-root  radius stalk-surface-above-ring stalk-surface-below-ring stalk-color-above-ring stalk-color-below-ring veil-type veil-color  weight ring-number ring-type spore-print-color population habitat
0         x           f         g       t    n               f            c         b          u           t          b      32                        s                        s                      p                      g         p          w      23           o         p                 k          y       d
1         x           y         g       f    f               f            c         b          g           e          b       7                        k                        k                      b                      b         p          w      23           o         l                 h          v       d
2         b           s         w       t    a               f            c         b          w           e          c       2                        s                        s                      w                      w         p          w      24           o         p                 k          s       m
3         f           y         y       f    f               f            c         b          p           e          b       2                        k                        k                      n                      b         p          w      53           o         l                 h          y       d
4         b           y         y       t    a               f            c         b          w           e          c      10                        s                        s                      w                      w         p          w      24           o         p                 k          n       m

In [30]:
test=test_original

# Preprocessing Test data

In [31]:
categorical_variable=test.dtypes[test.dtypes=='object'].index.values

In [32]:
independent_variable= [ x for x in test.columns if x not in [dependent_variable]]

In [33]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

for var in categorical_variable:
    test[var]=le.fit_transform(test[var])

In [34]:
#storing prediction array for test data
prediction2=model2.predict(test[independent_variable])

In [35]:
#test data class column= prediction 2
test['class']=prediction2

# Final processed test data of species

In [36]:
test.head()

cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  gill-spacing  gill-size  gill-color  stalk-shape  stalk-root  radius  stalk-surface-above-ring  stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  weight  ring-number  ring-type  spore-print-color  population  habitat  class
0          4            0          3        1     5                1             0          0           9            1           1      32                         2                         2                       6                       3          0           2      23            1          4                  2           5        0      0
1          4            2          3        0     2                1             0          0           2            0           1       7                         1                         1                       0                       0          0           2      23            1          2                  1           4        0      1
2          0            1          8        1     0                1             0          0          10            0           2       2                         2                         2                       7                       7          0           2      24            1          4                  2           3        3      0
3          1            2          9        0     2                1             0          0           7            0           1       2                         1                         1                       4                       0          0           2      53            1          2                  1           5        0      1
4          0            2          9        1     0                1             0          0          10            0           2      10                         2                         2                       7                       7          0           2      24            1          4                  2           2        3      0

<BR>
<br>
<br>
<br>
<BR>
<br>
<br>
<br>

# PART 2 [Not for Evaluation]

## First predicting missing stalk-root data and then creating model 

<br>

<br>

## First Segregating know data from unknown data

In [37]:
#unknown data (test data)
new_test_data=data[data['stalk-root']==0]


In [38]:
#known data (train data)
new_train_data=data[data['stalk-root'] !=0]


In [39]:
#splitting new_train_data
train, test=train_test_split(new_train_data, test_size=.2, stratify=new_train_data['stalk-root'])

# Decision Tree Classifier  on new train data

In [40]:
model3=DecisionTreeClassifier()

## Changing Independent and Dependent variable


In [41]:
dependent_variable= 'stalk-root'
independent_variable= [ x for x in data.columns if x not in [dependent_variable]]

In [42]:
# Training the model
model3.fit(train[independent_variable], train[dependent_variable])

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [43]:
#Prediction array
prediction3=model3.predict(test[independent_variable])
prediction3[:5]



array([1, 1, 1, 1, 2])

## Accuracy Score

In [44]:
accuracy_score(prediction3, test['stalk-root'])



1.0

<br><br>

# Completing new_test_data

<br><br>

In [45]:
new_test_data.head()


class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  gill-spacing  gill-size  gill-color  stalk-shape  stalk-root  radius  stalk-surface-above-ring  stalk-surface-below-ring  stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  weight  ring-number  ring-type  spore-print-color  population  habitat
0       1          2            2          2        0     7                1             0          1           0            1           0     215                         2                         1                       6                       7          0           2      38            1          0                  7           4        2
2       0          2            2          4        1     5                1             0          0           1            0           0     244                         2                         2                       2                       2          0           2      37            2          0                  7           1        6
3       1          5            2          2        0     8                1             0          1           0            1           0     286                         1                         2                       6                       7          0           2      82            1          0                  7           4        2
8       1          5            3          2        0     2                1             0          1           0            1           0      60                         2                         2                       6                       6          0           2      50            1          0                  7           4        4
10      1          5            2          4        0     8                1             0          1           0            1           0     385                         1                         2                       6                       6          0           2      34            1          0                  7           4        0

## Deletion of stalk-root to apply model 3 on new_test_data

In [46]:
del new_test_data['stalk-root']

In [47]:
print new_test_data.shape


(1727, 24)


In [48]:
prediction3=model3.predict(new_test_data)


In [49]:
# To prevent this Warning: A value is trying to be set on a copy of a slice from a DataFrame.
new_test_data.is_copy = False

In [50]:
new_test_data['stalk-root']=prediction3


# Combining original training data

In [51]:
frames = [new_train_data, new_train_data]
data = pd.concat(frames)

<br><br><br>

# Building model to predict 

In [52]:
train, test = train_test_split(data, test_size=0.2, stratify=data['class'])

In [53]:
#describing independent and dependent variables again
dependent_variable= 'class'
independent_variable= [ x for x in data.columns if x not in [dependent_variable]]

In [54]:
model3=LogisticRegression()

In [55]:
model3.fit(train[independent_variable], train[dependent_variable])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [56]:
prediction3= model3.predict(test[independent_variable])

In [57]:
accuracy_score(prediction3, test['class'])

0.97727272727272729

## Accuracy using Logistic regression increased !

This model however can't be applied to test data because in test data too stalk-root of some sets are not given

And to apply this model first we need missing stalk root data but it itself demands 'class' values. Hence model will not work

<br>